lib/elixir_search_extractor/file_upload.ex

```diff
@@ -0,0 +1,60 @@
+defmodule ElixirSearchExtractor.FileUpload do
+  import Ecto.Query, warn: false
+
+  alias ElixirSearchExtractor.FileUpload.CsvUploader
+  alias ElixirSearchExtractor.FileUpload.CsvValidator
+  alias ElixirSearchExtractor.FileUpload.KeywordFile
+  alias ElixirSearchExtractor.Repo
+
+  def list_keyword_files do
+    Repo.all(KeywordFile)
+  end
+
+  def create_keyword_file(attributes, user_id) do
+    csv_file = attributes["csv_file"]
+
+    with :ok <- CsvValidator.validate_file(csv_file),
+         {:ok, upload_file_path} <- CsvUploader.upload_file(user_id, csv_file),
+         {:ok, refactored_attributes} <-
+           refactor_attributes(attributes, user_id, upload_file_path),
+         {:ok, changeset} <- create_record(refactored_attributes) do
+      validate_changeset(changeset)
+    else
+      {:error, reason} -> {:error, reason}
+    end
+  end
```

Comment:
I have another idea, maybe we could use Multi, so we don't need to clean up the file 🤔 

```elixir
def create_and_upload_keyword_file_multi(attributes) do
    Multi.new()
    |> Multi.insert(:create_keyword_file, KeywordFile.changeset(attributes))
    |> Multi.run(:upload_keyword_file, fn _, _ ->
      CsvUploader.upload_file(csv_file, dest_path)
    end)
end

# Then this can be refactored to

with :ok <- CsvValidator.validate_file(csv_file),
     {:ok, %{create_keyword_file: keyword_file}} <- 
           Repo.transaction(create_and_upload_keyword_file_multi()) do
  {:ok, keyword_file}
else
  {:error, :create_keyword_file, changeset, _} ->
     {:error, changeset}

  {:error, :upload_keyword_file, reason, _} ->
     {:error, reason}

  {:error, reason} -> 
     {:error, reason}
end
```

So if the record cannot insert, the file won't be uploaded. And if file upload failed, the record insertion is rollbacked. But we will need to build the upload path before. what do you think?

Reply:
@rosle I did more refactor according to your suggestion [4f75815](https://github.com/rafayet-monon/elixir_search_extractor/pull/35/commits/4f758151c958bcdcaf2367709e46b1f004e1a6e7) 😄